web3/decoding.nim

```diff
@@ -0,0 +1,437 @@
+    when isDynamic(typeof(field)):
+      offsets[i] = decoder.decode(uint64)
+    else:
+      field = decoder.decode(typeof(field))
+    inc i
+
+  i = 0
+  # Decode dynamic fields using the offsets.
+  for field in res.fields:
+    when isDynamic(typeof(field)):
+      let pos = decoder.input.pos()
+      if offsets[i].int > pos:
+        decoder.input.advance(offsets[i].int - pos)
+      field = decoder.decode(typeof(field))
+
+    inc i
+
+  discard offsets
+
+  return res
+
+proc decodeObject(decoder: var AbiDecoder, T: type): T {.raises: [SerializationError].} =
+  ## When T is a object, ABI layout looks like the typle:
+  ## +----------------------------+
+  ## | static field 0 or offset   |  <-- 32
+  ## +----------------------------+
+  ## | static field 1 or offset   |  <-- 32
+  ## +----------------------------+
+  ## | ...                        |
+  ## +----------------------------+
+  ## | static field N-1 or offset |  <-- 32
+  ## +----------------------------+
+  ## | dynamic field 0 data       |  <-- at offset
+  ## +----------------------------+
+  ## | dynamic field 1 data       |  <-- at offset
+  ## +----------------------------+
+  ## | ...                        |
+  ## +----------------------------+
+  var resultObj: T
+  var offsets = newSeq[uint64](totalSerializedFields(T))
+
+  # Decode static fields first and get the offsets for dynamic fields.
+  var i = 0
+  resultObj.enumInstanceSerializedFields(_, fieldValue):
+    when isDynamic(typeof(fieldValue)):
+      offsets[i] = decoder.decode(uint64)
+    else:
+      fieldValue = decoder.decode(typeof(fieldValue))
+    inc i
+
+  # Decode dynamic fields using the offsets.
+  i = 0
+  resultObj.enumInstanceSerializedFields(_, fieldValue):
+    when isDynamic(typeof(fieldValue)):
+      let pos = decoder.input.pos()
+      if offsets[i].int > pos:
+        decoder.input.advance(offsets[i].int - pos)
+      fieldValue = decoder.decode(typeof(fieldValue))
+    inc i
+
+  resultObj
+
+proc decode*(decoder: var AbiDecoder, T: type): T {.raises: [SerializationError]} =
+  ## This method should not be used directly.
+  ## It is needed because `genFunction` create tuple
+  ## with object instead of creating a flat tuple with
+  ## object fields.
+  when T is object:
+    let value = decoder.decodeObject(T)
+    return value
+  else:
+    let value = decoder.decode(T)
+    decoder.finish()
+    return value
+
+proc readValue*[T](r: var AbiReader, value: T): T {.raises: [SerializationError]} =
+  try:
+    readValue[T](r, T)
+  except SerializationError as e:
+    raise newException(SerializationError, e.msg)
+
+proc readValue*[T](r: var AbiReader, _: typedesc[T]): T {.raises: [SerializationError]} =
+  var resultObj: T
+  var decoder = AbiDecoder(input: r.getStream)
+  type StInts = StInt | StUint
+
+  when T is object and T is not StInts:
+    resultObj = decodeObject(decoder, T)
+  else:
+    resultObj = decoder.decode(T)
+
+  decoder.finish()
+  result = resultObj
```

Comment:
```suggestion
  resultObj
```

broadly, we encourage expression return: https://status-im.github.io/nim-style-guide/language.result.html